src/geometry.nim

```diff
@@ -496,21 +465,74 @@ proc newRotZ*(angle: float32): Rotation =
     ]
 
 
+proc apply*[T](transf: Transformation, x: T): T =
+    case transf.kind
+    of tkIdentity: return x
+
+    of tkGeneric, tkComposition, tkRotation: 
+        when T is Point3D: 
+            return dot(transf.mat, x.toVec4).toPoint3D
+        elif T is Normal:
+            return dot(x, transf.inv_mat).toNormal
+        else: 
+            return dot(transf.mat, x) 
+
+    of tkTranslation: 
+        when T is Vec4f:
+            return newVec4(x[0] + transf.mat[0][3] * x[3], x[1] + transf.mat[1][3] * x[3], x[2] + transf.mat[2][3] * x[3], x[3])
+        elif T is Vec3f:
+            return newVec3(x[0] + transf.mat[0][3], x[1] + transf.mat[1][3], x[2] + transf.mat[2][3])
+        elif T is Point3D: 
+            return newPoint3D(x.x + transf.mat[0][3], x.y + transf.mat[1][3], x.z + transf.mat[2][3])
+        elif T is Normal:
+            return newNormal(x.x + transf.inv_mat[3][0], x.y + transf.inv_mat[3][1], x.z + transf.inv_mat[3][2])
+
+    of tkScaling:
+        when T is Vec4f:
+            return newVec4(x[0] * transf.mat[0][0], x[1] * transf.mat[1][1], x[2] * transf.mat[2][2], x[3])
+        elif T is Vec3f:
+            return newVec3(x[0] * transf.mat[0][0], x[1] * transf.mat[1][1], x[2] * transf.mat[2][2])
+        elif T is Point3D: 
+            return newPoint3D(x.x * transf.mat[0][0], x.y * transf.mat[1][1], x.z * transf.mat[2][2])
+        elif T is Normal: 
+            return x
+
+
+proc `*`*(transf: Transformation, scal: float32): Transformation {.inline.} = 
+    result.kind = if transf.kind == tkIdentity: tkGeneric else: transf.kind
```

Comment:
not tkGenerig but tkScaling, isn't it?

Reply:
No i can multiply also a tkTranslation or other kinds of transformation

Reply:
Yes but if transform kind is identity and you multiply it by a scalar you have a scaling no?